rfcs/XXXX-separate-wal.md

```diff
@@ -0,0 +1,113 @@
+they might be significantly more expensive or might be limited in features. Using
+such stores for the main store containing the entire database might be undesirable.
+
+## Goals
+
+- Support an optional separate object store dedicated specifically for WAL.
+- Provide a user-facing configuration mechanism.
+- Avoid significant refactorings.
+
+## Public API
+
+The only change on the public API level is configuration:
+
+```rust
+pub struct DbOptions {
+    // ...
+
+    /// An optional [object store](ObjectStore) dedicated specifically for WAL.
+    ///
+    /// If not set, the main object store passed to `Db::open(...)` will be used
+    /// for WAL storage.
+    ///
+    /// NOTE: WAL durability properties depend on the durability properties of
+    ///       the underlying object store. Make sure the configured object store
+    ///       is durable enough for your use case.
+    #[serde(skip)]
+    pub wal_object_store: Option<Arc<dyn ObjectStore>>
+}
+```
+
+By default `wal_object_store` is set to `None`.
+
+## Implementation
+
+The configured WAL object store is propagated down to `TableStore` instance
+which internally will select the appropriate object store for each operation
+based solely on the `SsTableId` involved in the operation. The object path
+structure will remain unchanged for both stores.
+
+There is already `transactional_wal_store` field in `TableStore`. The naming
+is misleading because apparently this store is used not only for WAL-related
+operations. The field should be renamed, the resulting field layout will look
+like this:
+
+```rust
+pub struct TableStore {
+    // ...
+
+    /// The main object store.
+    object_store: Arc<dyn ObjectStore>,
+    /// The dedicated WAL object store, if any.
+    wal_object_store: Option<Arc<dyn ObjectStore>>,
+
+    /// The transactional store wrapper of the main object store.
+    transactional_store: Arc<dyn TransactionalObjectStore>,
+    /// The transactional store wrapper of the dedicated WAL object store, if any.
+    wal_transactional_store: Option<Arc<dyn TransactionalObjectStore>>
+}
+```
+
+The object store selection procedure will look like this:
+
+```rust
+fn object_store_for(&self, id: &SsTableId) -> Arc<dyn ObjectStore> {
+    match id {
+        SsTableId::Wal(..) =>
+            if let Some(wal_object_store) = &self.wal_object_store {
+                wal_object_store.clone()
+            } else {
+                self.object_store.clone()
+            }
+        SsTableId::Compacted(..) => self.object_store.clone(),
+    }
+}
+```
+
+The transactional object store selection logic would look about the same.
+
+### Block Cache
+
+Interaction with the block cache is completely transparent. After a suitable
+object store is selected for the operation, no special care is needed to make it
+work with the block cache.
+
+### Object Cache
+
+Interaction with the on-disk object cache is more complex. To support caching,
```

Comment:
in #519, i might have a try about making a refactor to let us not reading WALs except replaying WALs on start up, with this change established, the benefits of having an object cache over wal objects will reduce a lot.

Reply:
this would be great

Reply:
https://github.com/slatedb/slatedb/pull/533/commits/ecc25cfa957e3ec590e385526d6a50903abc45f0